### PR TITLE
Fix target for doc test cross compilation

### PR DESCRIPTION
**Pull request by Freax13**
- for target paths `Compilation::target` only contains the shortened file name, but we need to pass the full path to rustdoc

### CODE REVIEW

**Comment by ehuss** [comment]
Thanks!

It seems like `Compilation.host` and `Compilation.target` only exist for doc tests. Perhaps those should just be removed?  Then `run_doc_tests` can just look at the `build_config` to decide what to do.

Also, can you add a test for this?  There are some examples in [`custom_target.rs`](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/custom_target.rs).  There is documentation for writing tests in [`CONTRIBUTING`](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) and [`cargo-test-support`](https://github.com/rust-lang/cargo/blob/master/crates/cargo-test-support/src/lib.rs).

**Comment by Freax13** [comment]
> It seems like `Compilation.host` and `Compilation.target` only exist for doc tests. Perhaps those should just be removed? Then `run_doc_tests` can just look at the `build_config` to decide what to do.

[`Compilation`](https://docs.rs/cargo/0.43.1/cargo/core/compiler/struct.Compilation.html) is part of the public api though so removing host and target is technically a breaking change

**Comment by Freax13** [comment]
> Also, can you add a test for this? There are some examples in [`custom_target.rs`](https://github.com/rust-lang/cargo/blob/master/tests/testsuite/custom_target.rs). There is documentation for writing tests in [`CONTRIBUTING`](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) and [`cargo-test-support`](https://github.com/rust-lang/cargo/blob/master/crates/cargo-test-support/src/lib.rs).

there is a similar bug [#71002](https://github.com/rust-lang/rust/pull/71002) in `rustdoc` that needs to get patched first. when that pr is merged I can add tests

**Comment by ehuss** [comment]
> [`Compilation`](https://docs.rs/cargo/0.43.1/cargo/core/compiler/struct.Compilation.html) is part of the public api though so removing host and target is technically a breaking change

Cargo's public API isn't stable, so each release is a breaking change.

**Comment by Freax13** [comment]
> It seems like `Compilation.host` and `Compilation.target` only exist for doc tests. Perhaps those should just be removed? Then `run_doc_tests` can just look at the `build_config` to decide what to do.

I played around with this and it seems to me that removing those two attribute results in messy/more complex code, so imho it's not worth removing them.

**Comment by ehuss** [comment]
Hm, ok, I see that doctests have the unusual behavior that `--target HOST` still needs to work.

Perhaps add a comment explaining that, perhaps along the lines of "Use `rustc_target()` to properly handle JSON spec paths.".

Let me know if you have any issues adding the test.

**Comment by Freax13** [comment]
> Let me know if you have any issues adding the test.

okay so the fix in rustdoc got merged this morning, so I'm guessing it should be in the next nightly version. I'll try adding the tests tomorrow (and the comment regarding `rustc_target`)

File: tests/testsuite/cross_compile.rs [hunk 1]
```diff
@@ -437,6 +437,45 @@ fn no_cross_doctests() {
...
+            //! ```
+        "#,
+        )
+        .build();
+
+    let target = cross_compile::alternate();
+
+    // This tests the library and runs the doc tests.
+    p.cargo("test -v -Z doctest-xcompile --target")
+        .arg(&target)
```
[the remark this answers is not in the document]
Comment (Freax13) [reply tests/testsuite/cross_compile.rs:462]: yeah that makes sense not sure why I missed that

**Comment by Freax13** [comment]
the custom targets obviously can't use std, so exiting cleanly is a bit of a challenge. the only way I can think of right now is by using inline assembly and manually calling the system calls to exit. the downside of this approach is that it wouldn't be cross platform

**Comment by ehuss** [comment]
Hm, I'm wondering, are you somehow using json specs to build std?  I'm having a hard time imagining an environment where these kinds of tests would even work.  (I'm only familiar with using specs with no_std.)

**Comment by Freax13** [comment]
I'm using it for my os kernel using the bootimage/bootloader crates which procss the compiled binaries into a disk image

**Comment by ehuss** [comment]
Can you say more of how that would work? Without std, I don't think the tests will work (maybe with a custom test runner, but those don't work with rustdoc afaik).

**Comment by Freax13** [comment]
correct. i'm using custom test frameworks for normal tests. doctests are actually just small programs that all get compiled and executed individually.

usually rustdoc wraps the code in a main method, but if it detects that you wrote your own, it will use that instead. that way you can also use outer attributes like #![no_std] or use extern crate.

so in my case I have some hidden code in the doctests that runs the os initialization and the os uses qemu's exit decice feature to stop itself and signal whether the tests ran successfully. if i remember correctly i think i managed to get it to work at one point by manually fixing all the target path bugs.
i checked some cross compiled binaries earlier and it seems like they were normal elf files (im not sure if thats the default or if it is influenced by some config key i set). those elf files also had an entrypoint, so i was able to open them in a debugger (they pretty much instantly segfault, but thats probably caused by my kernel trying to setup memory and not because they're inherintly broken)

**Comment by ehuss** [comment]
That is an impressive level of dedication to get doctests working! :smile:

Considering that getting an actual doctest working will likely be very difficult, perhaps it should just have a test that cargo passes the correct `--target` flag to rustdoc.  So maybe in `custom_target_minimal`, add something like the following:

```rust
    // Ensure that the correct style of flag is passed to --target with doc tests.
    p.cargo("test --doc --target src/../custom-target.json -v -Zdoctest-xcompile")
        .masquerade_as_nightly_cargo()
        .with_stderr_contains("[RUNNING] `rustdoc [..]--target [..]foo/custom-target.json[..]")
        .run();
```

with the goal just to check that it passes the correct form of `--target`. How does that sound?

**Comment by Freax13** [comment]
sounds good to me

**Comment by ehuss** [comment]
Thanks!
@bors r+

**Pull request merged**